### PR TITLE
Evo2 predict support for CP, TP and DP with new test coverage.

### PR DESCRIPTION
**Pull request by jstjohn**
### Description

Work through bugs with context parallelism through the prediction task for evo2. Demonstrate good results in the BRCA notebook.

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
from bionemo.testing.data.fasta import ALU_SEQUENCE, create_fasta_file
create_fasta_file("example_small_128k.fasta", 1, sequence_length=2**17, repeating_dna_pattern=ALU_SEQUENCE)
```

Note the nccl disable setting is to address a bug specific to some A6000 setups. It should not be used in general.
```bash
CKPT_PATH=$(download_bionemo_data evo2/1b-8k-bf16:1.0)
NCCL_P2P_DISABLE=1 predict_evo2 --context-parallel-size 2 --fasta example_small_128k.fasta --model-size 1b --write-interval batch --output-dir results_128k --ckpt-dir $CKPT_PATH --output-log-prob-seqs --log-prob-collapse-option sum 
```

### Pre-submit Checklist

- [x] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

## Summary by CodeRabbit

- New Features
  - Unified predictor interface with per-token or per-sequence log-prob outputs, seq_idx included in predictions, new PredictDataModule, and updated default model size.
  - PredictionWriter gains distributed-aware options (per-rank saving, files-per-subdir) and more flexible write intervals.

- Refactor
  - Consolidated prediction flow under a common BasePredictor; concrete predictor classes introduced and logging/shuffle warnings added.

- Tests
  - Expanded multi-GPU/TP/PP/FP8 coverage and log-prob equivalence checks; tests updated to DP-rank-aware filenames.

- Chores
  - NeMo submodule update; minor .gitignore tweak.

### CODE REVIEW

**Comment by jstjohn** [comment]
This PR should make it easier to switch to `--write-interval batch` which was brought up in #910

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py [hunk 1]
```diff
@@ -180,38 +203,60 @@ def predict_step(self, batch, batch_idx: int | None = None) -> Tensor:
...
             softmax_logprobs = softmax_logprobs[:, :-1]
-            input_ids = batch["tokens"][:, 1:]
-            assert softmax_logprobs.shape[1] == input_ids.shape[1]
+            input_ids = tokens_gathered[:, 1:]
+            try:
+                assert softmax_logprobs.shape[1] == input_ids.shape[1]
+            except Exception as e:
+                if torch.distributed.get_rank() == 0:
+                    breakpoint()
+                torch.distributed.barrier()
```
Comment (cspades) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py]: Leftover breakpoint 👀

File: sub-packages/bionemo-llm/src/bionemo/llm/utils/callbacks.py [hunk 2]
```diff
@@ -65,16 +85,38 @@ def setup(self, trainer: pl.Trainer, pl_module: pl.LightningModule, *args, **kwa
...
+        return torch.distributed.get_rank(parallel_state.get_data_parallel_group(with_context_parallel=False))
+
+    @property
+    def should_write_predictions(self) -> bool:
+        """Returns the context parallel rank."""
+        # TODO: handle expert parallelism and other kinds of parallelism
+        return self.save_all_model_parallel_ranks or (
+            parallel_state.is_pipeline_last_stage()
+            and parallel_state.get_tensor_model_parallel_rank() == 0
+            and parallel_state.get_context_parallel_rank() == 0
```
Comment (cspades) [inline sub-packages/bionemo-llm/src/bionemo/llm/utils/callbacks.py]: Did you ever get an answer if pipeline parallelism is or isn't orthogonal to the other parallelisms? If that were the case you can get away with something like:
```
parallel_state.is_pipeline_last_stage() and torch.distributed.get_rank() % (torch.distributed.get_world_size() / parallel_state.get_pipeline_model_parallel_world_size()) == 0
```
to generalize for all other possible dimensions of parallelism.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py [hunk 3]
```diff
@@ -144,17 +163,20 @@ def _gather_along_cp_dim(input_, seq_dim: int = 1):
     dim_size[0] = dim_size[0] * world_size
 
     output = torch.empty(dim_size, dtype=input_.dtype, device=torch.cuda.current_device())
+    # TODO: handle zigzag packing here. Currently this just gathers along ranks, but if you want to see the sequence in
+    #   the original order you need to undo the zigzag packing that happens in
+    #   `megatron.core.utils.get_batch_on_this_cp_rank`.
     torch.distributed.all_gather_into_tensor(
-        output, input_.contiguous(), group=parallel_state.get_tensor_model_parallel_group()
+        output, input_.contiguous(), group=parallel_state.get_context_parallel_group()
```
Comment (cspades) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py:180]: Was this the main idea of this PR? To migrate from context parallelism backed by the TP group in Hyena to the CP group?

**Review by cspades — APPROVED** [review]

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py [hunk 4]
```diff
@@ -18,8 +18,8 @@
 
 
 import argparse
+import logging
```
Comment (yzhang123) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py]: could you pls use "from nemo.utils import logging"

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py [hunk 5]
```diff
@@ -91,6 +105,11 @@ def parse_args():
         default=None,
         help="Output dir that will contain the generated text produced by the Evo2 model. If not provided, the output will be logged.",
     )
+    ap.add_argument(
```
Comment (yzhang123) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py:118]: could you add a check that this exists if writer-interval is batch??

File: sub-packages/bionemo-llm/src/bionemo/llm/utils/callbacks.py [hunk 6]
```diff
@@ -17,15 +17,21 @@
...
 import lightning.pytorch as pl
 import torch
 from lightning.pytorch.callbacks import BasePredictionWriter
+from megatron.core import parallel_state
 
 from bionemo.llm.lightning import batch_collator
 
 
 IntervalT = Literal["epoch", "batch"]
+logger = logging.getLogger(__name__)
```
Comment (yzhang123) [inline sub-packages/bionemo-llm/src/bionemo/llm/utils/callbacks.py]: user nemo logger?

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py (hunk 3 above)
Comment (jstjohn) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py:180]: No, this was just the wrong group to use for this function previously (`_gather_along_cp_dim`). Other changes had to be made as well.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py [hunk 7]
```diff
@@ -265,8 +310,10 @@ def hyena_predict_data_step(dataloader_iter) -> dict[str, torch.Tensor]:
 
     if parallel_state.is_pipeline_first_stage():
         required_device_keys.update(("tokens", "position_ids"))
+    include_seq_idx = False
     if parallel_state.is_pipeline_last_stage():
-        required_device_keys.update(("labels", "loss_mask", "seq_idx"))
+        include_seq_idx = True
+        required_device_keys.update(("labels", "loss_mask"))
```
[the remark this answers is not in the document]
Comment (jstjohn) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py]: This is only true for multi-token prediction.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py (hunk 5 above)
Comment (jstjohn) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/predict.py:118]: Sure, or maybe a warning? For testing or small runs it wouldn't matter.

**Review by yzhang123 — APPROVED** [review]

**Comment by jstjohn** [comment]
TODO include nemo version bump once https://github.com/NVIDIA-NeMo/NeMo/pull/14684 merges.

**Comment by dorotat-nv** [comment]
> Is it possible to add a unit test that tests those bugs?

@jstjohn ? what was this bug and how do we know it was fixed?

**Review by dorotat-nv — APPROVED** [review]

**Comment by jstjohn** [comment]
> > Is it possible to add a unit test that tests those bugs?
> 
> @jstjohn ? what was this bug and how do we know it was fixed?

@dorotat-nv it's covered in sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_predict.py in this PR but unfortunately since CI only has one GPU it wouldn't be tested regularly.

Ideally there would be a CI instance with 2 GPUs that support FP8. I ran these tests in computelab on a 2xb300 (has FP8) and locally on my 2xA6000 (no FP8). Note that the b300 instance I was using had 8xb300s and the test handles using a subset.

**Comment by trvachov** [comment]
👏
